forwarder/src/runtime/io.rs

```diff
@@ -0,0 +1,179 @@
+use crate::error::ContractError;
+use crate::runtime::sys::exports;
+use crate::runtime::{RegisterIndex, Runtime};
+use crate::types::Vec;
+use borsh::{BorshDeserialize, BorshSerialize};
+
+pub trait StorageIntermediate: Sized {
+    fn len(&self) -> usize;
+    fn is_empty(&self) -> bool;
+    fn copy_to_slice(&self, buffer: &mut [u8]);
+
+    fn to_vec(&self) -> Vec<u8> {
+        let len = self.len();
+        let mut buf = Vec::new();
```

Comment:
This pattern is used a couple of times across the patch and this is a minor suggestion, but it might be worth extracting it to a small function/method of an object.

```
        let mut buf = Vec::new();
        assert!(len <= buf.capacity());
        unsafe {
            buf.set_len(len);
        }
        ```